### PR TITLE
NO-TICKET - Clean Claiming

### PR DESCRIPTION
**Pull request by Davran331122**
## Jira Ticket

Clear claiming, and user tokens polling on Home Screen. Soon will be third part

### CODE REVIEW

File: app/src/main/AndroidManifest.xml [hunk 1]
```diff
@@ -29,7 +29,7 @@
         android:fullBackupContent="@xml/backup_rules"
         android:icon="@mipmap/ic_launcher"
         android:label="@string/app_name"
-        android:largeHeap="false"
+        android:largeHeap="true"
```
Comment (kabramovp2p) [inline app/src/main/AndroidManifest.xml:32]: -

File: app/src/main/java/org/p2p/wallet/home/ui/main/HomeElementItemMapper.kt [hunk 2]
```diff
@@ -28,15 +26,15 @@ class HomeElementItemMapper(
...
-        result += ethereumTokens.map { token ->
-            val claimStatus = ethereumBundleStatuses[token.publicKey]?.filterNotNull().orEmpty()
-            val canBeClaimed = claimStatus.all { status -> status.canBeClaimed() }
-            HomeElementItem.Claim(
-                token = token,
-                isClaimEnabled = canBeClaimed
-            )
-        }
+//
+//        result += ethereumTokens.map { token ->
```
Comment (kabramovp2p) [inline app/src/main/java/org/p2p/wallet/home/ui/main/HomeElementItemMapper.kt:30]: why commented?

**Review by kabramovp2p — APPROVED** [review]

File: app/src/main/java/org/p2p/wallet/home/ui/main/HomePresenter.kt [hunk 3]
```diff
@@ -103,22 +97,26 @@ class HomePresenter(
...
-        val username: Username? = null,
-        val areZerosHidden: Boolean,
-    ) {
-        val ethereumTokens: List<Token.Eth>
-            get() = ethereumState.ethereumTokens
-        val ethereumBundleStatuses: Map<String, List<ClaimStatus?>>
-            get() = ethereumState.ethereumBundleStatuses
+    override fun attach(view: HomeContract.View) {
+        super.attach(view)
+        launch {
```
Comment (gslevinkov) [inline app/src/main/java/org/p2p/wallet/home/ui/main/HomePresenter.kt:102]: could you extract launch body to a method?

File: app/src/main/java/org/p2p/wallet/home/ui/main/HomePresenter.kt [hunk 4]
```diff
@@ -103,22 +97,26 @@ class HomePresenter(
...
+            tokensPolling.shareTokenPollFlowIn(this).onEach {
+                val isRefreshing = it.first.isEmpty()
+                view.showRefreshing(isRefreshing)
+            }.collect {
+                val solTokens = it.first
```
Comment (gslevinkov) [inline app/src/main/java/org/p2p/wallet/home/ui/main/HomePresenter.kt:107]: val (solTokens, ethTokens) = it

File: app/src/main/java/org/p2p/wallet/home/ui/main/HomePresenter.kt [hunk 5]
```diff
@@ -103,22 +97,26 @@ class HomePresenter(
...
-        val areZerosHidden: Boolean,
-    ) {
-        val ethereumTokens: List<Token.Eth>
-            get() = ethereumState.ethereumTokens
-        val ethereumBundleStatuses: Map<String, List<ClaimStatus?>>
-            get() = ethereumState.ethereumBundleStatuses
+    override fun attach(view: HomeContract.View) {
+        super.attach(view)
+        launch {
+            tokensPolling.shareTokenPollFlowIn(this).onEach {
```
Comment (gslevinkov) [inline app/src/main/java/org/p2p/wallet/home/ui/main/HomePresenter.kt:103]: .onEach { solTokensToEthTokens ->

File: app/src/main/java/org/p2p/wallet/home/ui/main/UserTokensPolling.kt [hunk 6]
```diff
@@ -19,37 +29,68 @@ class UserTokensPolling(
...
-            Timber.d("Skipping tokens auto-update")
         }
     }
 
-    private suspend fun getEthereumState(): EthereumHomeState {
-        if (!ethAddressEnabledFeatureToggle.isFeatureEnabled) {
-            return EthereumHomeState()
+    private fun startPollingEthereum() {
+        launch {
+            if (ethAddressEnabledFeatureToggle.isFeatureEnabled) {
```
Comment (gslevinkov) [inline app/src/main/java/org/p2p/wallet/home/ui/main/UserTokensPolling.kt:80]: you can reduce indentation moving check for feature toggle in the beginning of the method

if (!ethAddressEnabledFeatureToggle.isFeatureEnabled){
    return 
}

File: app/src/main/java/org/p2p/wallet/home/ui/main/UserTokensPolling.kt [hunk 7]
```diff
@@ -19,37 +29,68 @@ class UserTokensPolling(
...
+                try {
+                    if (ethAddressEnabledFeatureToggle.isFeatureEnabled) {
```
Comment (gslevinkov) [inline app/src/main/java/org/p2p/wallet/home/ui/main/UserTokensPolling.kt:82]: you check again, there is no need for that

File: app/src/main/java/org/p2p/wallet/home/ui/main/UserTokensPolling.kt [hunk 8]
```diff
@@ -19,37 +29,68 @@ class UserTokensPolling(
...
+                        val bundles = ethereumInteractor.getListOfEthereumBundleStatuses()
+                        ethTokenProvider.launch(claimingTokens = bundles)
+                    }
+                } catch (e: CancellationException) {
+                    Timber.i("Cancelled tokens remote update")
+                } catch (e: Throwable) {
+                    Timber.e(e, "Failed polling tokens")
```
Comment (gslevinkov) [inline app/src/main/java/org/p2p/wallet/home/ui/main/UserTokensPolling.kt:89]: Failed polling ETH tokens

File: app/src/main/java/org/p2p/wallet/home/ui/main/models/ViewState.kt [hunk 9]
```diff
@@ -0,0 +1,13 @@
+package org.p2p.wallet.home.ui.main.models
+
+import org.p2p.core.token.Token
+import org.p2p.wallet.auth.model.Username
+import org.p2p.wallet.home.model.VisibilityState
+
+data class ViewState(
```
Comment (gslevinkov) [inline app/src/main/java/org/p2p/wallet/home/ui/main/models/ViewState.kt:7]: HomeFragmentViewState to remove ambiguity

File: ethereumkit/src/main/java/org/p2p/ethereumkit/external/EthereumModule.kt [hunk 10]
```diff
@@ -33,5 +34,6 @@ object EthereumModule {
                 gson = get(named(QUALIFIER_RPC_GSON))
             )
         }
+        single { EthereumTokensProvider(get(),get(),get()) }
```
Comment (gslevinkov) [inline ethereumkit/src/main/java/org/p2p/ethereumkit/external/EthereumModule.kt:37]: singleOf

File: ethereumkit/src/main/java/org/p2p/ethereumkit/external/core/CoroutineDispatchers.kt [hunk 11]
```diff
@@ -3,7 +3,7 @@ package org.p2p.ethereumkit.external.core
 import kotlinx.coroutines.CoroutineDispatcher
 import kotlinx.coroutines.Dispatchers
 
-internal interface CoroutineDispatchers {
+interface CoroutineDispatchers {
```
Comment (gslevinkov) [inline ethereumkit/src/main/java/org/p2p/ethereumkit/external/core/CoroutineDispatchers.kt:6]: conflict with org.p2p.wallet.infrastructure.dispatchers.CoroutineDispatchers

File: ethereumkit/src/main/java/org/p2p/ethereumkit/external/repository/EthereumKitRepository.kt [hunk 12]
```diff
@@ -68,80 +69,68 @@ internal class EthereumKitRepository(
...
-        ))
+    private suspend fun loadTokenBalances(
+        address: EthAddress,
+        tokenAddresses: List<EthAddress>,
+    ): List<TokenBalanceResponse> {
+        return balanceRepository.getTokenBalances(address, tokenAddresses).balances
     }
 
+    private suspend fun getMetadataAsync(tokenBalance: TokenBalanceResponse, contractAddress: EthAddress) =
+        coroutineScope {
```
Comment (gslevinkov) [inline ethereumkit/src/main/java/org/p2p/ethereumkit/external/repository/EthereumKitRepository.kt:126]: withContext doesn't work?

File: ethereumkit/src/main/java/org/p2p/ethereumkit/external/repository/EthereumTokensProvider.kt [hunk 13]
```diff
@@ -0,0 +1,80 @@
...
+        refreshJob?.cancel()
+        refreshJob = launch {
+            try {
+                delay(getDelayTimeInMillis(delayInMillis))
+                async { ethToken.emit(loadWallet()) }
+                async { erC20Tokens.emit(loadErc20Tokens(claimingTokens)) }
+            } catch (e: CancellationException) {
+                Timber.tag(TAG).d(e, "Cancellation of token fetching")
+            } catch (e: Throwable) {
+                Timber.tag(TAG).d(e, "Error on fetching eth tokens")
```
Comment (gslevinkov) [inline ethereumkit/src/main/java/org/p2p/ethereumkit/external/repository/EthereumTokensProvider.kt:50]: TImber.e?

**Review by gslevinkov — APPROVED** [review]

File: app/src/main/java/org/p2p/wallet/home/ui/main/HomePresenter.kt (hunk 4 above)
Comment (kabramovp2p) [reply app/src/main/java/org/p2p/wallet/home/ui/main/HomePresenter.kt:107]: can be just        }.collect { (solTokens, ethTokens) ->

**Pull request merged**